packages/grid/_modules_/grid/hooks/features/pagination/useGridPagination.ts

```diff
@@ -38,13 +38,10 @@ export const useGridPagination = (apiRef: GridApiRef): void => {
 
       setGridState((state) => ({
         ...state,
-        pagination: applyConstraints(
-          {
-            ...state.pagination,
-            page,
-          },
-          options.page,
```

Comment:
As far as I debugged through code I came to the conclusion that,

**1.** Initially on render, if the `page` prop is provided (controlled mode), the two effects run at:
https://github.com/mui-org/material-ui-x/blob/9953f25ca0f5d0c60777e2717dd8d11d4ea0249b/packages/grid/_modules_/grid/hooks/features/pagination/useGridPagination.ts#L86
https://github.com/mui-org/material-ui-x/blob/9953f25ca0f5d0c60777e2717dd8d11d4ea0249b/packages/grid/_modules_/grid/hooks/features/pagination/useGridPagination.ts#L111
and `applyConstraints` apply the pageProp at (line 23) through `options.prop`: 
https://github.com/mui-org/material-ui-x/blob/9953f25ca0f5d0c60777e2717dd8d11d4ea0249b/packages/grid/_modules_/grid/hooks/features/pagination/useGridPagination.ts#L16-L25

**2.** When the `page` prop is not provided (uncontrolled mode), it picks the `state.page` in `applyConstraints`. (line 23)

And when page is changed on click, the `setPage` is triggered and there it is setting the `page` of `GridPageChangeParams` from params at:
https://github.com/mui-org/material-ui-x/blob/9953f25ca0f5d0c60777e2717dd8d11d4ea0249b/packages/grid/_modules_/grid/hooks/features/pagination/useGridPagination.ts#L35-L48
(line 42-44) and thus no need to pass `options.page` at line 46. (for both uncontrolled and controlled mode)

Similarly for `setPageSize`, I think no need to pass `options.page`.

Reply:
From what I understood, the purpose of `applyConstraints` is to make sure that the new page number or new page size is within the limits. In both effects, the prop value should be used because in controlled mode the user could send an off-limit value. But, in `setPage` and `setPageSize`, the source of truth is the value in the argument, no prop should be considered.

---

I don't know if everyone agrees, but I think we should change the signature of `applyConstraints` to:

```diff
-function applyConstraints(state: GridPaginationState, pageProp?: number): GridPaginationState {
+function applyConstraints(state: GridPaginationState, newPage?: number, newPageSize?: number): GridPaginationState 
```
Where:
- `state` is always `state.pagination` with the old page and page size
- `newPage` is the new page number - it comes from the prop or the argument
- `newPageSize` is the new page size - it comes from the prop or the argument

The function returns the updated state with the new values clamped to the limits.

Reply:
> In both effects, the prop value should be used because in controlled mode the user could send an off-limit value. But, in setPage and setPageSize, the source of truth is the value in the argument, no prop should be considered.

Correct, this is what I also found out. And thus removed passing of  page prop`options.page`